### PR TITLE
R 3.1 double dispatch bugs

### PR DESCRIPTION
**Pull request by DavisVaughan**
This PR attempts to guard against all potential R 3.1 `lapply()` + dispatching bugs as mentioned in [this comment](https://github.com/r-lib/vctrs/issues/114#issuecomment-470294051).

### CODE REVIEW

**Comment by hadley** [comment]
Just to confirm, after these changes:

* what do packages that use vctrs need for R 3.1? Nothing?
* what do packages that extend vctrs need for R 3.1? Still `@importFrom vctrs vec_cast.integer`?

**Comment by DavisVaughan** [comment]
tl;dr) Even after this PR:

- On 3.1, interactive use of vctrs should always be done with a call to `library(vctrs)`.

- When building packages, for 3.1 you must `@importFrom vctrs vec_cast.<type>` for all types you might cast to.

---

Testing on rstudio.cloud with the current dev version of vctrs and R 3.1:

- Interactive use of calling `vctrs::vec_cast(iris$Species, iris$Species)` without `library(vctrs)` currently fails with the following error. This PR will _not_ fix this problem.

```r
Error in vec_cast.factor(iris$Species, iris$Species) : 
  could not find function "vec_cast.factor"
```

- If building a package and exporting a function like `wrap_cast()` below, then you get the same error as above when calling `wrap_cast(iris$Species, iris$Species)`. To fix this, you _must_ use `@importFrom vctrs vec_cast.factor`. Yes, this means that you must `@importFrom` _all_ of the exported `vec_cast.<type>` functions [as I do here](https://github.com/DavisVaughan/hardhat/blob/master/R/aaa.R#L16). This PR does _not_ fix this issue.

```r
wrap_cast <- function(x, to) {
  vctrs::vec_cast(x, to)
}
```

- In the scenario above with `wrap_cast()`, say you have done the imports below, and call `wrap_cast(iris, iris)`. You will still get the error below about not being able to find `vec_cast.double`. This PR _does_ solve this issue with the anonymous function fix.

```r
#' @importFrom vctrs vec_cast.data.frame
#' @importFrom vctrs vec_cast.double
```

```r
 Error in vec_cast.double(dots[[1L]][[1L]], dots[[2L]][[1L]]) : 
  could not find function "vec_cast.double" 
```

**Comment by lionel-** [comment]
Perhaps we should "fix" this by using internal genericity for base types. vctrs won't be extensible on R 3.1 but at least should work with base types.

**Comment by hadley** [comment]
@lionel- ah, I think that's a good point since we were intending to do that anyway for performance reasons (by internal genericity, I assume you're meaning switchspatch on `vctrs_type`)

**Comment by DavisVaughan** [comment]
So can this be closed because of this bit now?

https://github.com/r-lib/vctrs/blob/7da46b006134c81a0a3061d51846fd433c1abf26/R/zzz.R#L17

**Comment by lionel-** [comment]
> So can this be closed because of this bit now?

This was meant to fix another bug, but maybe it works for yours as well?

**Comment by DavisVaughan** [comment]
Yea it seems to have fixed it. I just reran my hardhat 3.1 travis build that had a Remote on vctrs and it is passing now. I think it's because `vec_cast()` itself no longer does the R level dispatch, and `vec_cast_dispatch()` does. Because `vec_cast_dispatch()` isn't called directly in the `lapply()`, it seems to work.

**Comment by hadley** [comment]
Ok, great. I'm going to close this and the related issue.
